File: MetaBrain/confidence/provider-consistency.ts
import type { ProviderConsistencyResult, ProviderOutputSummary } from "./types";

function normalize(value: string | undefined): string {
  return (value ?? "").toLowerCase().replace(/[^a-z0-9 ]/g, " ").replace(/\s+/g, " ").trim();
}

function similarity(left: string, right: string): number {
  if (!left && !right) return 0;
  if (left === right) return 1;
  const leftTerms = new Set(left.split(" ").filter(Boolean));
  const rightTerms = new Set(right.split(" ").filter(Boolean));
  const intersection = [...leftTerms].filter((term) => rightTerms.has(term)).length;
  const union = new Set([...leftTerms, ...rightTerms]).size || 1;
  return intersection / union;
}

export function evaluateProviderConsistency(outputs: ProviderOutputSummary[]): ProviderConsistencyResult {
  const usable = outputs.filter((output) => output.status === "ok" && normalize(output.content_summary));
  const unresolved: string[] = outputs
    .filter((output) => output.status !== "ok")
    .map((output) => `${output.provider_name}:${output.status}`);

  if (usable.length < 2) {
    return {
      providers_compared: usable.length,
      consistency_score: usable.length === 1 ? 0.7 : 0,
      conflicts_detected: unresolved.length > 0,
      dominant_provider: usable[0]?.provider_name,
      unresolved_conflicts: unresolved,
    };
  }

  const comparisons: number[] = [];
  for (let index = 0; index < usable.length; index += 1) {
    for (let next = index + 1; next < usable.length; next += 1) {
      comparisons.push(similarity(normalize(usable[index].content_summary), normalize(usable[next].content_summary)));
    }
  }

  const score = comparisons.reduce((sum, value) => sum + value, 0) / comparisons.length;
  if (score < 0.45) unresolved.push("provider_output_divergence");

  return {
    providers_compared: usable.length,
    consistency_score: Number(score.toFixed(3)),
    conflicts_detected: unresolved.length > 0,
    dominant_provider: usable.sort((a, b) => (b.confidence_score ?? 0) - (a.confidence_score ?? 0))[0]?.provider_name,
    unresolved_conflicts: unresolved,
  };
}

File: MetaBrain/confidence_py/provider_consistency.py
import re

from .types import ProviderConsistencyResult, ProviderOutputSummary


def _normalize(value: str | None) -> str:
    return re.sub(r"\s+", " ", re.sub(r"[^a-z0-9 ]", " ", (value or "").lower())).strip()


def _similarity(left: str, right: str) -> float:
    if not left and not right:
        return 0.0
    if left == right:
        return 1.0
    left_terms = set(left.split())
    right_terms = set(right.split())
    union = len(left_terms | right_terms) or 1
    return len(left_terms & right_terms) / union


def evaluate_provider_consistency(outputs: list[ProviderOutputSummary]) -> ProviderConsistencyResult:
    usable = [item for item in outputs if item.status == "ok" and _normalize(item.content_summary)]
    unresolved = [f"{item.provider_name}:{item.status}" for item in outputs if item.status != "ok"]

    if len(usable) < 2:
        return ProviderConsistencyResult(
            providers_compared=len(usable),
            consistency_score=0.7 if len(usable) == 1 else 0.0,
            conflicts_detected=bool(unresolved),
            dominant_provider=usable[0].provider_name if usable else None,
            unresolved_conflicts=unresolved,
        )

    comparisons: list[float] = []
    for index, output in enumerate(usable):
        for other in usable[index + 1 :]:
            comparisons.append(_similarity(_normalize(output.content_summary), _normalize(other.content_summary)))
    score = sum(comparisons) / len(comparisons)
    if score < 0.45:
        unresolved.append("provider_output_divergence")
    dominant = sorted(usable, key=lambda item: item.confidence_score or 0, reverse=True)[0]
    return ProviderConsistencyResult(
        providers_compared=len(usable),
        consistency_score=round(score, 3),
        conflicts_detected=bool(unresolved),
        dominant_provider=dominant.provider_name,
        unresolved_conflicts=unresolved,
    )
